ambry-clustermap/src/main/java/com/github/ambry/clustermap/AmbryPartitionStateModel.java

```diff
@@ -25,104 +26,188 @@
     }
   }
 
   @Transition(to = "DROPPED", from = "ERROR")
   public void onBecomeDroppedFromError(Message message, NotificationContext context) {
-    logger.info("Partition {} in resource {} is becoming DROPPED from ERROR", message.getPartitionName(),
-        message.getResourceName());
-    partitionStateChangeListener.onPartitionBecomeDroppedFromError(message.getPartitionName());
+    boolean shouldTransition = shouldTransition(message);
+    String partitionName = message.getPartitionName();
+    logger.info("Partition {} in resource {} is becoming DROPPED from ERROR, Should Transition? {}", partitionName,
+        message.getResourceName(), shouldTransition);
+    if (shouldTransition) {
+      partitionStateChangeListener.onPartitionBecomeDroppedFromError(partitionName);
+    }
   }
 
   @Transition(to = "OFFLINE", from = "ERROR")
   public void onBecomeOfflineFromError(Message message, NotificationContext context) {
-    logger.info("Partition {} in resource {} is becoming OFFLINE from ERROR", message.getPartitionName(),
-        message.getResourceName());
-    partitionStateChangeListener.onPartitionBecomeOfflineFromError(message.getPartitionName());
+    boolean shouldTransition = shouldTransition(message);
+    String partitionName = message.getPartitionName();
+    logger.info("Partition {} in resource {} is becoming OFFLINE from ERROR, Should Transition? {}", partitionName,
+        message.getResourceName(), shouldTransition);
+    if (shouldTransition) {
+      partitionStateChangeListener.onPartitionBecomeOfflineFromError(partitionName);
+    }
   }
 
   @Override
   public void reset() {
     logger.info("Reset method invoked. Partition {} in resource {} is reset to OFFLINE", partitionName, resourceName);
     partitionStateChangeListener.onReset(partitionName);
   }
+
+  /**
+   * Return true if we should transition the state from the message. Ambry is undergoing resource reconstruction. Same
+   * partition will be added to different resources. Resource that has higher id should be considered as the primary
+   * resource.
+   * @param message Message that carry partition information.
+   * @return True if a state transition should happen.
+   */
+  boolean shouldTransition(Message message) {
+    String resourceName = message.getResourceName();
+    String partitionName = message.getPartitionName();
+    String mappedResourceName = partitionNameToResourceName.compute(partitionName, (k, v) -> {
+      if (v == null || v.equals(resourceName)) {
+        return resourceName;
+      }
+      try {
+        int oldResourceId = Integer.valueOf(v);
+        int newResourceId = Integer.valueOf(resourceName);
+        return newResourceId > oldResourceId ? resourceName : v;
+      } catch (Exception e) {
+        logger.error("Failed to parse resource name to an integer", e);
+        throw new StateTransitionException("Failed to parse resource name",
```

Comment:
You might want to add the resourceName to see which resouceName failed to be parsed.

Reply:
The exception is NumberFormatException, which already have the input.